File: MensLacrosseStats/View Controllers/Game/INSOGameTableViewController.m
//
//  INSOGameTableViewController.m
//  ScorebookLite
//
//  Created by James Dabrowski on 9/24/15.
//  Copyright © 2015 Intangible Software. All rights reserved.
//

@import CoreData;

#import "INSOProductManager.h"
#import "MensLacrosseStatsAppDelegate.h"

#import "INSOGameTableViewController.h"
#import "INSOGameTableViewCell.h"
#import "INSOGameDetailViewController.h"

#import "Game.h"
#import "Event.h"
#import "RosterPlayer.h"

static NSString * const INSOGameCellIdentifier = @"GameCell";
static NSString * const INSOShowGameDetailSegueIdentifier = @"ShowGameDetailSegue";
static NSString * const INSOShowPurchaseModalSegueIdentifier = @"ShowPurchaseModalSegue";

@interface INSOGameTableViewController () <NSFetchedResultsControllerDelegate, INSOProductManagerDelegate>
// IBOutlets
@property (nonatomic, weak) IBOutlet UIBarButtonItem* addButton;

// IBActions
- (IBAction)addGame:(id)sender;

// Private Properties
@property (nonatomic) NSFetchedResultsController* gamesFRC;
@property (nonatomic) NSManagedObjectContext* managedObjectContext;

// Private Methods
- (void)configureTableView;
- (void)configureGameCell:(INSOGameTableViewCell*)cell atIndexPath:(NSIndexPath*)indexPath;

- (void)prepareForShowGameDetailSegue:(UIStoryboardSegue*)segue sender:(INSOGameTableViewCell*)cell;

@end

@implementation INSOGameTableViewController
#pragma mark - Lifecycle
- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.addButton.enabled = NO; 
    
    [INSOProductManager sharedManager].delegate = self;
    [[INSOProductManager sharedManager] refreshProduct];
    
    [self configureTableView];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

- (void)dealloc
{
    [INSOProductManager sharedManager].delegate = nil;
}

#pragma mark - IBActions
- (void)addGame:(id)sender
{
    if ([[INSOProductManager sharedManager] productIsPurchased] && ![[INSOProductManager sharedManager] productPurchaseExpired]) {
        // If the app is purchased, just create a new game. No hoo-hoo.
        [self createNewGame];
    } else {
        // Now we gotta see how many games there are.
        // If there is one already, don't let 'em create a new one.
        // Instead, show the buy me! modal.
        NSInteger numberOfGames = [[[self.gamesFRC sections] objectAtIndex:0] numberOfObjects];
        if (numberOfGames >= 1) {
            [self performSegueWithIdentifier:INSOShowPurchaseModalSegueIdentifier sender:self];
        } else {
            [self createNewGame]; 
        }
    }
}

#pragma mark - Private Properties
- (NSFetchedResultsController*)gamesFRC
{
    if (!_gamesFRC) {
        NSFetchRequest* request = [NSFetchRequest fetchRequestWithEntityName:[Game entityName]];
        [request setFetchBatchSize:50];
        
        NSSortDescriptor* sortByDate = [NSSortDescriptor sortDescriptorWithKey:@"gameDateTime" ascending:NO];
        [request setSortDescriptors:@[sortByDate]];
        
        _gamesFRC = [[NSFetchedResultsController alloc] initWithFetchRequest:request  managedObjectContext:self.managedObjectContext sectionNameKeyPath:nil cacheName:nil];
        _gamesFRC.delegate = self;
        
        NSError* error = nil;
        if (![_gamesFRC performFetch:&error]) {
            // Error fetching games
            NSLog(@"Error fetching games: %@", error.localizedDescription);
        }
    }
    
    return _gamesFRC;
}
                     
- (NSManagedObjectContext*)managedObjectContext
{
    if (!_managedObjectContext) {
        MensLacrosseStatsAppDelegate* appDelegate = (MensLacrosseStatsAppDelegate *)[[UIApplication sharedApplication] delegate];
        _managedObjectContext = appDelegate.managedObjectContext;
    }
    return _managedObjectContext;
}

#pragma mark - Private Methods
- (void)configureTableView
{
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    self.tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
    self.tableView.alwaysBounceVertical = NO;
    self.tableView.estimatedRowHeight = 189.0;
}

- (void)configureGameCell:(INSOGameTableViewCell *)cell atIndexPath:(NSIndexPath *)indexPath
{
    Game* game = [self.gamesFRC objectAtIndexPath:indexPath];

    NSString* dateFormat = [NSDateFormatter dateFormatFromTemplate:@"Mdyy" options:0 locale:[NSLocale currentLocale]];
    NSString* timeFormat = [NSDateFormatter dateFormatFromTemplate:@"hmma" options:0 locale:[NSLocale currentLocale]];
    NSString* dateTimeFormat = [NSString stringWithFormat:@"%@' at '%@", dateFormat, timeFormat];
    
    NSDateFormatter* formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat:dateTimeFormat];
    cell.gameDateTimeLabel.text = [formatter stringFromDate:game.gameDateTime];
    
    cell.homeTeamLabel.text = game.homeTeam;
    cell.homeScoreLabel.text = [NSString stringWithFormat:@"%@", game.homeScore];
    
    cell.visitingTeamLabel.text = game.visitingTeam;
    cell.visitingScoreLabel.text = [NSString stringWithFormat:@"%@", game.visitorScore];
    
    cell.locationLabel.text = game.location;
}

- (NSDate*)newGameStartDateTime
{
    NSDate* currentDate = [NSDate date];
    NSDateComponents* components = [[NSCalendar currentCalendar] components: (NSCalendarUnitDay | NSCalendarUnitMonth | NSCalendarUnitYear | NSCalendarUnitHour | NSCalendarUnitMinute) fromDate:currentDate];
    NSInteger minutes = components.minute;
    float minuteUnit = ceil((float) minutes / 30.0);
    minutes = minuteUnit * 30;
    [components setMinute:minutes];
    return [[NSCalendar currentCalendar] dateFromComponents:components];
}

- (void)createNewGame
{
    // Create a game with now as the game date time
    Game* newGame = [Game insertInManagedObjectContext:self.managedObjectContext];
    newGame.gameDateTime = [self newGameStartDateTime];
    
    // Team to record
    newGame.teamWatching = newGame.homeTeam;
    
    // Set up events to record
    NSArray* defaultEvents = [Event fetchDefaultEvents:self.managedObjectContext];
    NSSet* eventSet = [NSSet setWithArray:defaultEvents];
    [newGame addEventsToRecord:eventSet];
    
    // Give the game 2 team players
    RosterPlayer* teamPlayer = [RosterPlayer insertInManagedObjectContext:self.managedObjectContext];
    teamPlayer.numberValue = INSOTeamWatchingPlayerNumber;
    teamPlayer.isTeamValue = YES;
    [newGame addPlayersObject:teamPlayer];
    
    RosterPlayer *otherTeamPlayer = [RosterPlayer insertInManagedObjectContext:self.managedObjectContext];
    otherTeamPlayer.numberValue = INSOOtherTeamPlayerNumber;
    otherTeamPlayer.isTeamValue = YES;
    [newGame addPlayersObject:otherTeamPlayer];
    
    NSError* error = nil;
    if (![self.managedObjectContext save:&error]) {
        NSLog(@"Error saving MOC after creating new game: %@", error.localizedDescription);
    }
}

#pragma mark - Navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:INSOShowGameDetailSegueIdentifier]) {
        [self prepareForShowGameDetailSegue:segue sender:sender];
    }
    
    if ([segue.identifier isEqualToString:INSOShowPurchaseModalSegueIdentifier]) {
        [self prepareForShowPurchaseModalSegue:segue sender:sender];
    }
}

- (void)prepareForShowGameDetailSegue:(UIStoryboardSegue *)segue sender:(INSOGameTableViewCell *)cell
{
    NSIndexPath* indexPath = [self.tableView indexPathForCell:cell];
    Game* selectedGame = [self.gamesFRC objectAtIndexPath:indexPath];
    INSOGameDetailViewController* dest = segue.destinationViewController;
    dest.game = selectedGame; 
}

- (void)prepareForShowPurchaseModalSegue:(UIStoryboardSegue*)segue sender:(id)sender
{
    // Nothing to do here, yet.
}

#pragma mark - Table view data source
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [[[self.gamesFRC sections] objectAtIndex:section] numberOfObjects];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    INSOGameTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:INSOGameCellIdentifier forIndexPath:indexPath];
    
    // Configure the cell...
    [self configureGameCell:cell atIndexPath:indexPath];
    
    return cell;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        Game* game = [self.gamesFRC objectAtIndexPath:indexPath];
        [self.managedObjectContext deleteObject:game];
    }
    
    NSError* error = nil;
    if (![self.managedObjectContext save:&error]) {
        NSLog(@"Error saving MOC after deleting game: %@", error.localizedDescription);
    }
}

- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 189.0;
}

#pragma mark - NSFetchedResultsControllerDelegate
- (void)controllerWillChangeContent:(NSFetchedResultsController *)controller
{
    [self.tableView beginUpdates];
}

- (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)anObject atIndexPath:(NSIndexPath *)indexPath forChangeType:(NSFetchedResultsChangeType)type newIndexPath:(NSIndexPath *)newIndexPath
{
    switch (type) {
        case NSFetchedResultsChangeInsert:
            [self.tableView insertRowsAtIndexPaths:@[newIndexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
            break;
        case NSFetchedResultsChangeDelete:
            [self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
            break;
        case NSFetchedResultsChangeUpdate:
            [self.tableView reloadRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationNone];
        case NSFetchedResultsChangeMove:
            [self.tableView reloadData]; 
        default:
            break;
    }
}

- (void)controllerDidChangeContent:(NSFetchedResultsController *)controller
{
    [self.tableView endUpdates];
}

#pragma mark - INSOProductManagerDelegate
- (void)didRefreshProduct
{
    self.addButton.enabled = YES;
}

@end

File: MensLacrosseStats/View Controllers/Game/INSOGameTableViewController.swift
//  Converted to Swift 5.5 by Swiftify v5.5.27463 - https://swiftify.com/
//
//  INSOGameTableViewController.swift
//  ScorebookLite
//
//  Created by James Dabrowski on 9/24/15.
//  Copyright © 2015 Intangible Software. All rights reserved.
//

import CoreData
import UIKit

private let INSOGameCellIdentifier = "GameCell"
private let INSOShowGameDetailSegueIdentifier = "ShowGameDetailSegue"
private let INSOShowPurchaseModalSegueIdentifier = "ShowPurchaseModalSegue"

class INSOGameTableViewController: UITableViewController, NSFetchedResultsControllerDelegate {
    // Private Properties

    private lazy var gamesFRC: NSFetchedResultsController<Game> = {
        let request = NSFetchRequest<Game>(entityName: Game.entityName())
        request.fetchBatchSize = 50

        let sortByDate = NSSortDescriptor(key: "gameDateTime", ascending: false)
        request.sortDescriptors = [sortByDate]

        let gamesFRC = NSFetchedResultsController(fetchRequest: request, managedObjectContext: managedObjectContext, sectionNameKeyPath: nil, cacheName: nil)
        gamesFRC.delegate = self

        var error: Error? = nil
        do {
            try gamesFRC.performFetch()
        } catch {
            // Error fetching games
            print("Error fetching games: \(error.localizedDescription)")
        }

        return gamesFRC
    }()

    private var managedObjectContext: NSManagedObjectContext = {
        let appDelegate = UIApplication.shared.delegate as! AppDelegate
        return appDelegate.managedObjectContext
    }()

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()

        configureTableView()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    // MARK: - IBActions

    @IBAction func addGame(_ sender: Any?) {
        createNewGame()
    }

    // MARK: - Private Methods

    private func configureTableView() {
        tableView.rowHeight = UITableView.automaticDimension
        tableView.tableFooterView = UIView(frame: CGRect.zero)
        tableView.alwaysBounceVertical = false
        tableView.estimatedRowHeight = 189.0
    }

    private func configureGameCell(_ cell: INSOGameTableViewCell?, at indexPath: IndexPath?) {
        var game: Game? = nil
        if let indexPath = indexPath {
            game = gamesFRC.object(at: indexPath) as Game
        }

        let dateFormat = DateFormatter.dateFormat(fromTemplate: "Mdyy", options: 0, locale: NSLocale.current)
        let timeFormat = DateFormatter.dateFormat(fromTemplate: "hmma", options: 0, locale: NSLocale.current)
        let dateTimeFormat = "\(dateFormat ?? "")' at '\(timeFormat ?? "")"

        let formatter = DateFormatter()
        formatter.dateFormat = dateTimeFormat
        if let gameDateTime = game?.gameDateTime {
            cell?.gameDateTimeLabel.text = formatter.string(from: gameDateTime)
        }

        cell?.homeTeamLabel.text = game?.homeTeam
        if let homeScore = game?.homeScore {
            cell?.homeScoreLabel.text = "\(homeScore)"
        }

        cell?.visitingTeamLabel.text = game?.visitingTeam
        if let visitorScore = game?.visitorScore {
            cell?.visitingScoreLabel.text = "\(visitorScore)"
        }

        cell?.locationLabel.text = game?.location
    }

    func newGameStartDateTime() -> Date? {
        let currentDate = Date()
        var components = Calendar.current.dateComponents([.day, .month, .year, .hour, .minute], from: currentDate)
        var minutes = components.minute
        let minuteUnit = ceil(Float(minutes ?? 0) / 30.0)
        minutes = Int(minuteUnit * 30)
        components.minute = minutes
        return Calendar.current.date(from: components)
    }

    func createNewGame() {
        // Create a game with now as the game date time
        let newGame = Game.insert(in: managedObjectContext)
        newGame.gameDateTime = newGameStartDateTime()

        // Team to record
        newGame.teamWatching = newGame.homeTeam

        // Set up events to record
        let defaultEvents: [Event] = Event.fetchDefaultEvents(managedObjectContext) as! [Event]
        let eventSet: Set<Event> = Set(defaultEvents.map({ $0 }))
        newGame.addEvents(toRecord: eventSet)

        // Give the game 2 team players
        let teamPlayer = RosterPlayer.insert(in: managedObjectContext)
        teamPlayer.numberValue = Int16(INSOTeamWatchingPlayerNumber)
        teamPlayer.isTeamValue = true
        newGame.addPlayersObject(teamPlayer)

        let otherTeamPlayer = RosterPlayer.insert(in: managedObjectContext)
        otherTeamPlayer.numberValue = Int16(INSOOtherTeamPlayerNumber)
        otherTeamPlayer.isTeamValue = true
        newGame.addPlayersObject(otherTeamPlayer)

        var _: Error? = nil
        do {
            try managedObjectContext.save()
        } catch {
            print("Error saving MOC after creating new game: \(error.localizedDescription)")
        }
    }

    // MARK: - Navigation

    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // If for some reason the sender is not our game table view cell, just don't do anything.
        guard let sender = sender as? INSOGameTableViewCell else {
            return
        }
        
        // Check the identifier, be safe.
        if segue.identifier == INSOShowGameDetailSegueIdentifier {
            prepare(forShowGameDetailSegue: segue, sender: sender)
        }
    }

    private func prepare(forShowGameDetailSegue segue: UIStoryboardSegue?, sender cell: INSOGameTableViewCell) {
        // If for some reason we don't have an index path, time to just bail.
        guard let indexPath = tableView.indexPath(for: cell) else {
            return
        }
        
        let selectedGame: Game? = gamesFRC.object(at: indexPath) as Game
        let dest = segue?.destination as? INSOGameDetailViewController
        dest?.game = selectedGame
    }

    // MARK: - Table view data source

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return gamesFRC.sections?[section].numberOfObjects ?? 0
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: INSOGameCellIdentifier, for: indexPath) as! INSOGameTableViewCell

        // Configure the cell...
        configureGameCell(cell, at: indexPath)

        return cell
    }

    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        // Return NO if you do not want the specified item to be editable.
        return true
    }

    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            if let game = gamesFRC.object(at: indexPath) as Game? {
                managedObjectContext.delete(game)
            }
        }

        var _: Error? = nil
        do {
            try managedObjectContext.save()
        } catch {
            print("Error saving MOC after deleting game: \(error.localizedDescription)")
        }
    }

    override func tableView(_ tableView: UITableView, estimatedHeightForRowAt indexPath: IndexPath) -> CGFloat {
        return 189.0
    }

    // MARK: - NSFetchedResultsControllerDelegate

    func controllerWillChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        tableView.beginUpdates()
    }

    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange anObject: Any, at indexPath: IndexPath?, for type: NSFetchedResultsChangeType, newIndexPath: IndexPath?) {
        switch type {
        case .insert:
            tableView.insertRows(at: [newIndexPath].compactMap { $0 }, with: .automatic)
        case .delete:
            tableView.deleteRows(at: [indexPath].compactMap { $0 }, with: .automatic)
        case .update:
            tableView.reloadRows(at: [indexPath].compactMap { $0 }, with: .none)
        case .move:
            tableView.reloadData()
        default:
            break
        }
    }

    func controllerDidChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        tableView.endUpdates()
    }
}
